### PR TITLE
feat: add /health_check

### LINKED ISSUE #16: Endpoint | Health check
Create a heartbeat/heath check endpoint in order to check the status of the service.

### PR DESCRIPTION
**Pull request by morenol**
This PR adds a health_check endpoint. Right now we are only testing the connection with the database.

 Fixes https://github.com/rust-lang-ve/hilow/issues/16

### CODE REVIEW

File: src/handlers/health_check.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    let db_status = value == 1;
+
+    let status = ServiceStatus::new(db_status);
+    HttpResponse::Ok().json(status)
+}
+
+/// Service Health Check
+#[derive(Serialize)]
+struct ServiceStatus {
```
Comment (morenol) [inline src/handlers/health_check.rs:23]: Where should we put this struct? @rust-lang-ve/hilow

File: src/state.rs [hunk 2]
```diff
@@ -29,18 +30,19 @@ pub struct State {
 impl State {
     /// Create a new state with the given database_url
     pub async fn new(database_url: &str) -> anyhow::Result<Self> {
-        let db = PgPool::connect(database_url).await?;
+        let db = PgPoolOptions::new()
+            .connect_timeout(std::time::Duration::from_secs(2))
```
Comment (JMEspiz) [inline src/state.rs:34]: Maybe it could be a good idea to define the DB timeout limit using an Env Var, what do you think?

Comment (morenol) [reply src/state.rs]: Done!

**Comment by EstebanBorai** [comment]
Perhaps is a good idea merging this after #21?

File: src/handlers/health_check.rs (hunk 1 above)
Comment (EstebanBorai) [reply src/handlers/health_check.rs:23]: I think the right place for this is here as this is not used in any other place. Lets land this before the other PR, I will solve conflicts and update them

**Review by LeoBorai — APPROVED** [review]

File: src/domain/models/profile.rs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+use uuid::Uuid;
+
+pub struct Profile {
+    pub id: Uuid,
+    pub user_id: Uuid,
+    pub avatar_id: Option<Uuid>,
+    pub name: String,
+    pub surname: Option<String>,
```
Comment (rafacouto) [inline src/domain/models/profile.rs:8]: Commented before about 'surname'

File: migrations/20210313213657_create_profiles.sql [hunk 4]
```diff
@@ -0,0 +1,10 @@
+-- Add migration script here
+CREATE TABLE IF NOT EXISTS profiles (
+  id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+  name VARCHAR(120) NOT NULL,
+  surname VARCHAR(120),
```
[the remark this answers is not in the document]
Comment (morenol) [reply migrations/20210313213657_create_profiles.sql:5]: Hi @rafacouto please review this in https://github.com/rust-lang-ve/hilow/pull/21. This PR is on of these changes.

File: README.md [hunk 5]
```diff
@@ -34,13 +34,19 @@
...
+
 Key | Description | Example
 --- | --- | ---
 `RUST_LOG` | Log level | `warn,info,error,debug`
 `POSTGRES_USER` | PostgreSQL User | N/A
 `POSTGRES_PASSWORD` | PostgreSQL Password | N/A
 `POSTGRES_DB` | PostgreSQL Database Name | N/A
 `POSTGRES_HOST` | PostgreSQL Hostname | N/A
+`POSTGRES_CONN_TIMEOUT` | PostgreSQL connection timeout in seconds | 2
+`DATABASE_URL` | Database URL used by SQLx | `postgres://user:secret@127.0.0.1/dbname`
```
[the remark this answers is not in the document]
Comment (morenol) [reply README.md:48]: sqlx uses DATABASE_URL and the POSTGRES docker container uses the other ones. I think that we can use https://github.com/dotenv-rs/dotenv#variable-substitution to construct DATABASE_URL from the  POSTGRES_ by default in the .env.sample file.

cc @EstebanBorai

[the remark this answers is not in the document]
Comment (rafacouto) [reply README.md:48]: +1 to build DATABASE_URL from POSTGRES variables

[the remark this answers is not in the document]
Comment (EstebanBorai) [reply README.md:48]: If it works with SQLx as its working now Im in!

[the remark this answers is not in the document]
Comment (EstebanBorai) [reply README.md:48]: @morenol is right here, this suggestion is out of scope.

Also the definition for the entity was done before at the definition of the problem.

We use `surname` because you will need to search for users Name and Surname separately to improve search results.

You can have:

- Email
- Username
- Name
- Surname

To find profiles around the app.

File: src/bin/main.rs [hunk 6]
```diff
@@ -1,8 +1,4 @@
-use actix_web::web::Data;
-use actix_web::{App, HttpServer};
-use anyhow::Context;
-use hilow::router::router;
-use hilow::state::State;
+use hilow::run::run;
```
Comment (EstebanBorai) [inline src/bin/main.rs:1]: I think we could have a `server` module for this instead, `run` doesn't mean too much.
What do you think about defining a `server` module and have it export this function there?

So this would be:

```rust
use hilow::server::run
```

File: src/database.rs [hunk 7]
```diff
@@ -20,8 +20,13 @@ pub fn make_db_pool() -> Result<DbPool> {
         .parse::<u32>()
         .context("Invalid value provided for \"DB_MAX_CONNECTIONS\" expected numeric value")?;
 
+    let connection_timeout = match var("POSTGRES_CONN_TIMEOUT") {
+        Ok(timeout) => timeout.parse().unwrap_or(2),
+        Err(_) => 2,
+    };
```
Comment (EstebanBorai) [inline src/database.rs:26]: This looks like a nice use case for `unwrap_or_else`:

```suggestion
    let connection_timeout = var("POSTGRES_CONN_TIMEOUT")
        .unwrap_or_else(|_| String::from("2"))
        .parse:<u32>()
        .context("The value provided to \"POSTGRES_CONN_TIMEOUT\" is not numeric")?;
```

File: src/handlers/health_check.rs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+use actix_web::{web, HttpResponse};
```
Comment (EstebanBorai) [inline src/handlers/health_check.rs:1]: This module (`handlers`) could live inside of the `server` module suggested above

File: src/router.rs [hunk 9]
```diff
@@ -1,8 +1,9 @@
-use actix_web::web::{post, resource, ServiceConfig};
```
Comment (EstebanBorai) [inline src/router.rs:1]: This module (`router`) would be a nice candidate for the `server` module suggested above

File: src/run.rs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+use actix_web::{dev::Server, web::Data, App, HttpServer};
+use anyhow::{Context, Result};
+
+use log::info;
+
+use crate::router::router;
+use crate::state::State;
+
+pub fn run(address: &str) -> Result<Server> {
```
Comment (EstebanBorai) [inline src/run.rs:9]: Please make use of `SocketAddr` for this, lets avoind using primitive data types when we have dedicated types available.

This way we would have parameter validation by design.

Refer to: https://doc.rust-lang.org/std/net/enum.SocketAddr.html

File: src/state.rs [hunk 11]
```diff
@@ -1,6 +1,9 @@
-use anyhow::Result;
 use std::sync::Arc;
 
+use anyhow::Result;
+
```
Comment (EstebanBorai) [inline src/state.rs:4]: I think you leave this extra line behind?

File: src/state.rs [hunk 12]
```diff
@@ -9,17 +12,17 @@ use crate::domain::services::Services;
...
+    pub db_pool: Arc<PgPool>,
 }
 
 impl State {
     pub fn new() -> Result<Self> {
         let db_pool = make_db_pool()?;
         let db_pool = Arc::new(db_pool);
 
-        let services = Services::new(db_pool)?;
+        let services = Services::new(db_pool.clone())?;
```
Comment (EstebanBorai) [inline src/state.rs:23]: Pleas use associated function `clone` from `Arc` here:

```suggestion
        let services = Services::new(Arc::clone(&db_pool))?;
```

File: tests/health_check.rs [hunk 13]
```diff
@@ -0,0 +1,36 @@
+use hilow;
+#[actix_rt::test]
```
Comment (EstebanBorai) [inline tests/health_check.rs:2]: Please leave a line between the `use` statement and the `#[actix_rt::test]` annotation

File: src/bin/main.rs [hunk 14]
```diff
@@ -12,20 +8,6 @@ async fn main() -> anyhow::Result<()> {
...
-        // `Data<State>` is being cloned to avoid creating
-        // one `State` instance per worker.
-        // Otherwise, `Data<State>` will be intialized as many
-        // times as the total count of workers configured
-        App::new().app_data(state.clone()).configure(router)
-    })
-    .bind("0.0.0.0:7878")?
-    .run()
-    .await?;
+    run("0.0.0.0:7878")?.await?;
```
Comment (EstebanBorai) [inline src/bin/main.rs:11]: If the suggestion above is applied, then we would have more context on what this function really does, having something similar to:

```rust
server::run("0.0.0.0:7878".parse())?.await?;
```

**Review by LeoBorai — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 15]
```diff
@@ -23,4 +29,6 @@ sqlx = { version = "0.4", features = ["offline", "postgres", "runtime-actix-nati
 uuid = "0.8"
 
 [dev-dependencies]
-actix-rt = "2"
+actix-rt = "1"
```
[the remark this answers is not in the document]
Comment (morenol) [reply Cargo.toml:32]: actix-rt is compatible with tokio version 1. 

Most of the other dependencies that we are using are compatible with previous versions of tokio. This caused me problems in the tests (in this PR we are adding the usage o reqwest and tokio in the tests).

You can run `cargo tree` in the code that we currently have in main.

File: src/run.rs (hunk 10 above)
Comment (morenol) [reply src/run.rs]: I changed to TcpListener, that allows us to use random ports to spawn the application in the integration tests

File: src/server/run.rs [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    let server = HttpServer::new(move || {
+        // `Data<State>` is being cloned to avoid creating
+        // one `State` instance per worker.
+        // Otherwise, `Data<State>` will be intialized as many
+        // times as the total count of workers configured
+        App::new().app_data(state.clone()).configure(router)
+    })
+    .listen(listener)?
+    .run();
+    Ok(server)
```
Comment (EstebanBorai) [inline src/server/run.rs:25]: Whats the benefit on providing a `TcpListener` ourselves? Is there anything we get by doing this approach?

Actix just needs the desired socket to bind to, this is not something you usually do, using a `SocketAddr` is just enough to achieve this. If we wanted to do something else with this TCP connection apart from using it with Actix perhaps this would help, but at least for now theres no need to consume the same TCP stream as Actix.

If validation is you goal using `SocketAddr` here will do the job, Actix is also capable of receiving this `SocketAddr` instance and bind the server to the corresponding host as you expect.

**Review by LeoBorai — CHANGES_REQUESTED** [review]

File: src/bin/main.rs [hunk 17]
```diff
@@ -11,21 +9,9 @@ async fn main() -> anyhow::Result<()> {
     }
 
     env_logger::init();
+    let listener =
+        TcpListener::bind("0.0.0.0:7878").expect("Unable to bind the service in 0.0.0.0:7878");
```
[the remark this answers is not in the document]
Comment (morenol) [reply src/bin/main.rs:13]: Yes, It if fine for the integration tests, we can use TcpListener to use random ports to spawn the servers in the tests

File: tests/health_check.rs [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+        .await
+        .expect("Failed to  execute request.");
+
+    assert!(response.status().is_server_error());
+}
+
+/// Spawn a hilow server and return its address to the caller.
+fn spawn_app() -> Result<String> {
+    // Port 0 translates to a random port.
+    let listener = TcpListener::bind("127.0.0.1:0").expect("Failed to bind random port");
```
[the remark this answers is not in the document]
Comment (morenol) [reply tests/health_check.rs:40]: Because if we use 7878, we cannot run tests locally while we have the process running. The port `0` will make that the OS looks for a random available port and use that. I don't see how this could make this function to fail.

File: tests/health_check.rs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    // We retrieve the port assigned to us by the OS
+    let port = listener.local_addr().unwrap().port();
+    let server = hilow::server::run(listener).unwrap();
+    let _ = tokio::spawn(server);
```
[the remark this answers is not in the document]
Comment (morenol) [reply tests/health_check.rs:44]: Yes, but this is only done in the integration tests in order to spawn several instances of the server.

File: src/server/handlers/health_check.rs [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+use serde::Serialize;
+use sqlx::postgres::PgRow;
+use sqlx::Row;
+
+use crate::state::State;
+
+pub async fn health_check(state: web::Data<State>) -> HttpResponse {
+    let value = sqlx::query("SELECT 1")
+        .try_map(|row: PgRow| row.try_get::<i32, _>(0))
+        .fetch_one(&*state.db_pool)
```
[the remark this answers is not in the document]
Comment (morenol) [reply src/server/handlers/health_check.rs]: It was failing when I tried that because &Arc<Pool> does not implement Executor. I fixed this by using:

`state.db_pool.as_ref()`

File: src/bin/main.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply src/bin/main.rs:13]: I don't think this is the way to solve that. If we were intercepting this TCP stream someway by piping the stream for instance, creating our own TCP connection would make sense.

The issue here is strongly related to configuration and tests. What you want to avoid is having conflicts with OS sockets between the server and your tests suite.

This approach is affecting the way the server is implemented due to a rare case where the developer wants to run test while running the server.

You can easily fix this by relying on configuration and leave the server startup simple, clean and using just the minimal resources required.

I suggest defining environment variables with default values for this.

You would end up having predefined ports por both cases, server and tests, and you will be able to define them in the same place as all of the other configurations

[the remark this answers is not in the document]
Comment (morenol) [reply src/bin/main.rs:13]: The problem is that I want a way to spawn servers in the integration tests with random ports, if I just decide to define a PORT from the environment for the tests, then we could have the "Address already in use" due to different tests are using the same port.

File: Cargo.toml (hunk 15 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply Cargo.toml:32]: Understood!

File: tests/health_check.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply tests/health_check.rs:40]: Is there a need to run our tests in parallel? Are you having any trouble running this tests with the previous server implementation?

File: tests/health_check.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply tests/health_check.rs:44]: Why we want to spawn several instances of the server?

**Review by LeoBorai — CHANGES_REQUESTED** [review]

File: tests/health_check.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (morenol) [reply tests/health_check.rs:44]: one per each test

**Comment by EstebanBorai** [comment]
@morenol theres something in regards to running multiple instances of the server in parallel during tests I think I'm not aware of, if you don't mind elaborating a bit on that would help me, perhaps I'm missing something

File: tests/health_check.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (morenol) [reply tests/health_check.rs:40]: no, but even with --test-threads=1 we can have problems if we use the same port for each test.

File: src/server/handlers/health_check.rs [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+pub async fn health_check(state: web::Data<State>) -> HttpResponse {
+    let db_status = sqlx::query("SELECT 1")
+        .fetch_one(state.db_pool.as_ref())
+        .await
+        .is_ok();
+    let service_status = ServiceStatus::new(db_status);
+
+    service_status.into()
+}
```
[the remark this answers is not in the document]
Comment (morenol) [reply src/server/handlers/health_check.rs:14]: I like this way, I just made it in the another way to make it easier to change at the moment when we will add more checks, like the redis check.

File: tests/health_check.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply tests/health_check.rs:44]: But why we would do that? I don't see that pattern here: https://actix.rs/docs/testing/
Actix handles that internally, usually you don't run isolated instances of your server AFAIK.

Many HTTP frameworks provide lightweight solutions for testing where you can run tests against them and these frameworks handles the way the connections are done. Why not following the Actix documentation for this?

Usually you run: `test::init_service(App::new()..`, what's the benefit behind this granular control? And why we need parallel instance during tests?

File: src/server/handlers/health_check.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply src/server/handlers/health_check.rs:14]: Yes but when integrating Redis we would probably have access to the driver via `Data<State>` and we will be able to do something similar to: `state.redis.ping()`

**Comment by morenol** [comment]
> @morenol theres something in regards to running multiple instances of the server in parallel during tests I think I'm not aware of, if you don't mind elaborating a bit on that, perhaps I'm missing something

The idea is to create an instance of the server for each test in order to isolate the feature that I want to test. Think about those tests as integration tests

File: tests/health_check.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply tests/health_check.rs:40]: Regularly each test in Rust runs on its own thread but as we are using Actix Runtime which runs on Tokio, even if you decide to consume 1 thread Tokio is able to spawn N green threads (Tokio tasks) in the same thread.

Actually you shouldn't have to worry about this because Actix provides its own runtime which is used to solve this issues.

Please read on: https://docs.rs/actix-rt/2.1.0/actix_rt/index.html for more details, but we should be spawning this using Tokio ourselves, we just have to mark our test functions with `actix_rt::test` and leave Actix Runtime handle this, we should focus on write our tests and make them pass.

These tests should look as straightforward as:

```rust
    #[actix_rt::test]
    async fn test_index_get() {
        let mut app = test::init_service(App::new().route("/", web::get().to(index))).await;
        let req = test::TestRequest::with_header("content-type", "text/plain").to_request();
        let resp = test::call_service(&mut app, req).await;
        assert!(resp.status().is_success());
    }
```

[Reference](https://actix.rs/docs/testing/)

**Comment by EstebanBorai** [comment]
> > @morenol theres something in regards to running multiple instances of the server in parallel during tests I think I'm not aware of, if you don't mind elaborating a bit on that, perhaps I'm missing something
> 
> The idea is to create an instance of the server for each test in order to isolate the feature that I want to test. Think about those tests as integration tests

Thing is that by doing this approach we are going over some optimizations an issues Actix Runtime already have solved on top of Tokio as well. We should rely on Actix Runtime to accomplish this ([here](https://docs.rs/actix-rt/2.1.0/actix_rt/index.html)) which already takes care of this and have support for unit and integration tests as well.

Actix Runtime acts as a proxy to Tokio runtime and takes care of futures emitted during our tests as well:
https://github.com/actix/actix-net/blob/master/actix-rt/src/system.rs

By spawning our own Tokio task we have to handle all of these edge cases ourselves.

Here is how Actix recommends how tests must be implemented: https://actix.rs/docs/testing/

File: tests/health_check.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (morenol) [reply tests/health_check.rs:44]: I saw that here https://github.com/LukeMathWalker/zero-to-production/blob/main/tests/api/health_check.rs

I will take a look at those documentation, I wasnt aware about them

**Comment by morenol** [comment]
> Thing is that by doing this approach we are going over some optimizations an issues Actix Runtime already have solved on top of Tokio as well. We should rely on Actix Runtime to accomplish this ([here](https://docs.rs/actix-rt/2.1.0/actix_rt/index.html)) which already takes care of this and have support for unit and integration tests as well.

Thanks, I will take a look to those links!

File: Cargo.toml (hunk 15 above)
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply Cargo.toml:32]: Is still required to use `actix-rt` on version `1` now that we are using the same crate for testing?

File: src/bin/main.rs [hunk 22]
```diff
@@ -11,21 +7,6 @@ async fn main() -> anyhow::Result<()> {
...
-        // `Data<State>` is being cloned to avoid creating
-        // one `State` instance per worker.
-        // Otherwise, `Data<State>` will be intialized as many
-        // times as the total count of workers configured
-        App::new().app_data(state.clone()).configure(router)
-    })
-    .bind("0.0.0.0:7878")?
-    .run()
-    .await?;
+    run("0.0.0.0:7878").await?;
```
Comment (EstebanBorai) [inline src/bin/main.rs:10]: What do you think about:

1. Take HOST or use a default from environment variables
2. Take PORT or use a default from environment variables
3. Create an instance of `std::net::SocketAddr` with provided or default HOST/PORT environment variables

Thing is that I think its better to rely on types than primitives for values that require validation. Using a `SocketAddr` here is more semantic and also give us validation OOTB for this use case

File: src/server/handlers/health_check.rs [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+        Ok(_) => HttpResponse::Ok().json(ServiceStatus { db_status: true }),
+        Err(_) => HttpResponse::InternalServerError().json(ServiceStatus { db_status: false }),
+    }
+}
+
+/// Service Health Check
+#[derive(Serialize)]
+struct ServiceStatus {
+    db_status: bool,
+}
```
Comment (EstebanBorai) [inline src/server/handlers/health_check.rs:20]: Lets avoid use before implementations in our files, that makes reading the code more natural without the need of scrolling to find definitions.

I suggest moving this struct definition before its usage in `health_check` function.

This struct is only being used as a response here, so why not appending `Response` word at the end? `HealthCheckResponse` would be more semantic, WDYT?

**Review by LeoBorai — CHANGES_REQUESTED** [review]

File: src/bin/main.rs (hunk 22 above)
Comment (morenol) [reply src/bin/main.rs:10]: Yes, I agree on that, but in any case I think that it could be done in another PR, we can create an Issue for that

File: Cargo.toml (hunk 15 above)
[the remark this answers is not in the document]
Comment (morenol) [reply Cargo.toml:32]: I think that we can use actix-rt version 2 and all the tests will pass without problems but maybe we could find problems in the future if we keep using that version, what do you think?

File: src/bin/main.rs (hunk 22 above)
Comment (EstebanBorai) [reply src/bin/main.rs:10]: SGTM!

**Review by LeoBorai — APPROVED** [review]
LGTM

**Comment by EstebanBorai** [comment]
With #26 we should be able to merge this, @morenol
